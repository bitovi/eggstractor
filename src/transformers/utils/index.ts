export * from './deduplicate-messages.utils';
export * from './get-style-property-and-value.utils';
export * from './group-by.utils';
export * from './update-padding-based-on-border.utils';
