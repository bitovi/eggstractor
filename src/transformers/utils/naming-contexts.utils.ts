export interface NamingContext {
  env: 'css' | 'scss ' | 'tailwind-v4' | 'tailwind-v3-sass';
  prefix?: string;
  includePageInPath?: boolean;
  delimiters: {
    pathSeparator: string;
    afterComponentName: string;
    variantEqualSign: string;
    betweenVariants: string;
  };
  duplicate?: (name: string, count: number) => string;
}

export const defaultContext: NamingContext = {
  env: 'css',
  includePageInPath: true,
  delimiters: {
    pathSeparator: '-',
    afterComponentName: '-',
    variantEqualSign: '_',
    betweenVariants: '-',
  },
  duplicate: (name, count) => `${name}${count}`,
};

export const tailwind4NamingConvention: NamingContext = {
  env: 'tailwind-v4',
  delimiters: {
    pathSeparator: '/',
    afterComponentName: '.',
    variantEqualSign: '_',
    betweenVariants: '.',
  },
  duplicate: (name, count) => `${name}${count}`,
};
