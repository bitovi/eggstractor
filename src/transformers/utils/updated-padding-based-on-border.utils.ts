const BORDER_PROPERTIES = [
  'border',
  'border-width',
  'border-left-width',
  'border-right-width',
  'border-top-width',
  'border-bottom-width',
] as const;

const PADDING_PROPERTIES = [
  'padding',
  'padding-left',
  'padding-right',
  'padding-top',
  'padding-bottom',
] as const;

type CssUnit = 'px' | 'rem' | 'scss' | 'unknown' | 'none';

type ParsedCssValue =
  | {
      format: 'px' | 'rem' | 'none';
      value: number;
      text: string;
    }
  | {
      value: null;
      format: Exclude<CssUnit, 'px' | 'rem' | 'none'>;
      text: string;
    };

function parseCssValue(text: string): ParsedCssValue {
  // special case: "0" (no unit needed in CSS)
  if (text === '0' || text === '0px' || text === '0rem') {
    return { value: 0, text: '0', format: 'none' };
  }

  if (text.startsWith('$')) {
    return { value: null, text: `#{${text}}`, format: 'scss' };
  }

  // match numbers with px/rem
  const match = /^(-?\d*\.?\d+)(px|rem)?$/.exec(text);

  if (match) {
    const [, num, unit] = match;
    return {
      value: parseFloat(num),
      text,
      format: unit as 'px' | 'rem',
    };
  }

  // fallback
  return { value: null, text, format: 'unknown' };
}

/**
 * Subtracts border value from padding value, returning a valid CSS string.
 * Padding can never be negative, so the result is clamped at 0.
 */
function subtract(padding: string, border: string): string {
  const a = parseCssValue(padding);
  const b = parseCssValue(border);

  // first is zero → result is always "0"
  // second is zero → return the original first string as-is
  if (a.value === 0 || b.value === 0) {
    return a.text;
  }

  // same known units → numeric subtraction, clamped at 0
  if (
    a.format === b.format &&
    (a.format === 'px' || a.format === 'rem') &&
    (b.format === 'px' || b.format === 'rem')
  ) {
    const result = Math.max(0, a.value - b.value);
    return result === 0 ? '0' : `${result}${a.format}`;
  }

  // mismatched/unknown units → calc with clamp (no trimming)
  return `calc(${a.text} - ${b.text})`;
}

interface Padding {
  'padding-top': string;
  'padding-right': string;
  'padding-bottom': string;
  'padding-left': string;
}

const parsePadding = (padding: string): Padding => {
  const parts = padding.trim().split(/\s+/);

  switch (parts.length) {
    case 1:
      return {
        'padding-top': parts[0],
        'padding-right': parts[0],
        'padding-bottom': parts[0],
        'padding-left': parts[0],
      };
    case 2:
      return {
        'padding-top': parts[0],
        'padding-right': parts[1],
        'padding-bottom': parts[0],
        'padding-left': parts[1],
      };
    case 3:
      return {
        'padding-top': parts[0],
        'padding-right': parts[1],
        'padding-bottom': parts[2],
        'padding-left': parts[1],
      };
    case 4:
      return {
        'padding-top': parts[0],
        'padding-right': parts[1],
        'padding-bottom': parts[2],
        'padding-left': parts[3],
      };
    default:
      throw new Error('Invalid padding shorthand');
  }
};

function toPaddingShorthand(paddings: Partial<Padding>): string {
  // Default missing sides to "0"
  const top = paddings['padding-top'] ?? '0';
  const right = paddings['padding-right'] ?? '0';
  const bottom = paddings['padding-bottom'] ?? '0';
  const left = paddings['padding-left'] ?? '0';

  // Case 1: all equal
  if (top === right && right === bottom && bottom === left) {
    return top;
  }

  // Case 2: top/bottom same, left/right same
  if (top === bottom && right === left) {
    return `${top} ${right}`;
  }

  // Case 3: left/right same
  if (right === left) {
    return `${top} ${right} ${bottom}`;
  }

  // Case 4: all different
  return `${top} ${right} ${bottom} ${left}`;
}

const hasBorderAndPaddingStyles = <T extends { styles: Record<string, string> }>(
  instance: T,
): boolean => {
  const hasPadding = PADDING_PROPERTIES.some((prop) => prop in instance.styles);

  if (!hasPadding) {
    return false;
  }

  return BORDER_PROPERTIES.some((prop) => prop in instance.styles);
};

const parseBorderWidth = (borderStyleValue: string): string => {
  return borderStyleValue.split(' ')[0] ?? null;
};

export const updatePaddingStylesBasedOnBorder = <
  T extends {
    styles: {
      [K in (typeof BORDER_PROPERTIES)[number] | (typeof PADDING_PROPERTIES)[number]]?: string;
    };
  },
>(
  instance: T,
): T => {
  if (!hasBorderAndPaddingStyles(instance)) {
    return instance;
  }

  const borderStyle = instance.styles['border']
    ? parseBorderWidth(instance.styles['border'])
    : null;

  const borders = {
    top: instance.styles['border-top-width'] || instance.styles['border-width'] || borderStyle,
    right: instance.styles['border-right-width'] || instance.styles['border-width'] || borderStyle,
    bottom:
      instance.styles['border-bottom-width'] || instance.styles['border-width'] || borderStyle,
    left: instance.styles['border-left-width'] || instance.styles['border-width'] || borderStyle,
  };

  const paddingStyles = instance.styles['padding']
    ? parsePadding(instance.styles['padding'])
    : null;

  const paddings = {
    'padding-top': instance.styles['padding-top'] || paddingStyles?.['padding-top'] || null,
    'padding-right': instance.styles['padding-right'] || paddingStyles?.['padding-right'] || null,
    'padding-bottom':
      instance.styles['padding-bottom'] || paddingStyles?.['padding-bottom'] || null,
    'padding-left': instance.styles['padding-left'] || paddingStyles?.['padding-left'] || null,
  };

  // Only add padding properties if they exist
  const newPaddingIndividualValues: { [K in (typeof PADDING_PROPERTIES)[number]]?: string } = {};

  if (paddings['padding-top']) {
    newPaddingIndividualValues['padding-top'] = borders.top
      ? subtract(paddings['padding-top'], borders.top)
      : paddings['padding-top'];
  }

  if (paddings['padding-right']) {
    newPaddingIndividualValues['padding-right'] = borders.right
      ? subtract(paddings['padding-right'], borders.right)
      : paddings['padding-right'];
  }

  if (paddings['padding-bottom']) {
    newPaddingIndividualValues['padding-bottom'] = borders.bottom
      ? subtract(paddings['padding-bottom'], borders.bottom)
      : paddings['padding-bottom'];
  }

  if (paddings['padding-left']) {
    newPaddingIndividualValues['padding-left'] = borders.left
      ? subtract(paddings['padding-left'], borders.left)
      : paddings['padding-left'];
  }

  // If there was padding shorthand, set it as well
  const newPaddingShorthand = paddingStyles ? toPaddingShorthand(newPaddingIndividualValues) : null;

  // Only add individual or shorthand padding properties if they exist
  const newPaddingStyles: { [K in (typeof PADDING_PROPERTIES)[number]]?: string } = {};

  if (instance.styles['padding'] && newPaddingShorthand) {
    newPaddingStyles['padding'] = newPaddingShorthand;
  }

  if (instance.styles['padding-top'] && newPaddingIndividualValues['padding-top']) {
    newPaddingStyles['padding-top'] = newPaddingIndividualValues['padding-top'];
  }

  if (instance.styles['padding-right'] && newPaddingIndividualValues['padding-right']) {
    newPaddingStyles['padding-right'] = newPaddingIndividualValues['padding-right'];
  }

  if (instance.styles['padding-bottom'] && newPaddingIndividualValues['padding-bottom']) {
    newPaddingStyles['padding-bottom'] = newPaddingIndividualValues['padding-bottom'];
  }

  if (instance.styles['padding-left'] && newPaddingIndividualValues['padding-left']) {
    newPaddingStyles['padding-left'] = newPaddingIndividualValues['padding-left'];
  }

  return {
    ...instance,
    styles: {
      ...instance.styles,
      ...newPaddingStyles,
    },
  };
};
