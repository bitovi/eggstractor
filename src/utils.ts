export default {
  toBase64(str: string): string {
    const base64chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    const utf8str = encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
      (_, p1) => String.fromCharCode(parseInt(p1, 16))
    );
    let i = 0;
    let result = '';
    while (i < utf8str.length) {
      const char1 = utf8str.charCodeAt(i++);
      const char2 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const char3 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const enc1 = char1 >> 2;
      const enc2 = ((char1 & 3) << 4) | (char2 >> 4);
      const enc3 = ((char2 & 15) << 2) | (char3 >> 6);
      const enc4 = char3 & 63;
      result += base64chars[enc1] + base64chars[enc2] +
        (isNaN(char2) ? '=' : base64chars[enc3]) +
        (isNaN(char3) ? '=' : base64chars[enc4]);
    }
    return result;
  },
  sanitizeName(name: string): string {
    return name.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  sanitizeSegment(segment: string): string {
    return segment.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  rgbToHex(r: number, g: number, b: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    return '#' + rHex + gHex + bHex;
  },
  rgbaToString(r: number, g: number, b: number, a: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    const aHex = Math.round(a * 255).toString(16).padStart(2, '0');
    
    return a === 1 ? 
      '#' + rHex + gHex + bHex : 
      '#' + rHex + gHex + bHex + aHex;
  },
  groupBy<T>(arr: T[], key: (item: T) => string): Record<string, T[]> {
    return arr.reduce((groups, item) => {
      const k = key(item);
      if (!groups[k]) groups[k] = [];
      groups[k].push(item);
      return groups;
    }, {} as Record<string, T[]>);
  }
};