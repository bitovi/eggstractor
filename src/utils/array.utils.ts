export function groupBy<T>(arr: T[], key: (item: T) => string): Record<string, T[]> {
  return arr.reduce(
    (groups, item) => {
      const k = key(item);
      if (!groups[k]) groups[k] = [];
      groups[k].push(item);
      return groups;
    },
    {} as Record<string, T[]>,
  );
}
