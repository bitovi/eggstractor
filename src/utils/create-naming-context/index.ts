export * from './create-naming-context.utils';
export * from './naming-context.utils';
