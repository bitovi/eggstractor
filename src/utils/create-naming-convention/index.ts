export * from './create-naming-convention.utils';
export * from './naming-contexts.utils';
