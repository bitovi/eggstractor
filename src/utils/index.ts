export * from './color.utils';
export * from './create-naming-convention'
export * from './constants.utils';
export * from './delay.utils';
export * from './get-valid-stylesheet-format.utils';
export * from './node.utils';
export * from './string.utils';
export * from './test.utils';
export * from './units.utils';
export * from './value.utils';
