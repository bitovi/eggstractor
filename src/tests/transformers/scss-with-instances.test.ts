import { transformToScssWithInstances } from '../../transformers/scss.transformer';
import { collectTokens } from '../../services/collection.service';
import { createTestData } from '../../utils/test.utils';
import testData from '../fixtures/figma-test-data_instances.json';

describe('Instance Override SCSS Generation', () => {
  it('should generate instance mixins with overrides', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    // Mock the getNodeByIdAsync to return proper nodes
    const originalGetNodeById = figma.getNodeByIdAsync;
    figma.getNodeByIdAsync = jest.fn().mockImplementation(async (id: string) => {
      // Find the node in our test data
      const findNode = (nodes: any[], targetId: string): any => {
        for (const node of nodes) {
          if (node.id === targetId) {
            return node;
          }
          if (node.children) {
            const found = findNode(node.children, targetId);
            if (found) return found;
          }
        }
        return null;
      };
      
      return findNode([testData], id);
    });

    try {
      const tokens = await collectTokens(jest.fn());
      const result = await transformToScssWithInstances(tokens);

      // Should contain regular SCSS output
      expect(result.result).toContain('// Generated SCSS Variables');
      expect(result.result).toContain('// Generated SCSS Mixins');
      
      // Should contain instance-specific mixins
      expect(result.result).toContain('// Generated Instance Mixins (Override Styles Only)');
      
      // Should contain mixins for our test instances
      expect(result.result).toContain('@mixin primary-button');
      expect(result.result).toContain('@mixin secondary-button');
      
      // Instance mixins should only contain override styles
      const instanceMixinSection = result.result.split('// Generated Instance Mixins (Override Styles Only)')[1];
      
      if (instanceMixinSection) {
        // Primary button should have red background override (different from component's blue)
        expect(instanceMixinSection).toMatch(/@mixin primary-button[\s\S]*background.*#cc3333/);
        
        // Secondary button should have gray background override
        expect(instanceMixinSection).toMatch(/@mixin secondary-button[\s\S]*background.*#e6e6e6/);
      }

      expect(result.warnings).toBeDefined();
      expect(result.errors).toBeDefined();
    } finally {
      figma.getNodeByIdAsync = originalGetNodeById;
    }
  });

  it('should handle instances with no overrides', async () => {
    // Create test data with an instance that has no overrides
    const noOverrideData = {
      ...testData,
      children: [
        {
          ...testData.children[0],
          children: [
            testData.children[0].children[0], // Component set
            {
              "id": "instance-no-override",
              "name": "Unchanged Button",
              "type": "INSTANCE",
              "mainComponent": {
                "id": "component-1",
                "type": "COMPONENT"
              },
              "variantProperties": {
                "State": "Default"
              },
              // Same properties as component - no overrides
              "fills": [
                {
                  "type": "SOLID",
                  "color": {
                    "r": 0.2,
                    "g": 0.4,
                    "b": 0.8
                  },
                  "opacity": 1,
                  "visible": true
                }
              ],
              "strokeWeight": 1,
              "cornerRadius": 8,
              "width": 120,
              "height": 40
            }
          ]
        }
      ]
    };

    const { setupTest } = createTestData(noOverrideData);
    const testSetup = await setupTest();
    global.figma = testSetup.figma;

    const originalGetNodeById = figma.getNodeByIdAsync;
    figma.getNodeByIdAsync = jest.fn().mockImplementation(async (id: string) => {
      const findNode = (nodes: any[], targetId: string): any => {
        for (const node of nodes) {
          if (node.id === targetId) return node;
          if (node.children) {
            const found = findNode(node.children, targetId);
            if (found) return found;
          }
        }
        return null;
      };
      return findNode([noOverrideData], id);
    });

    try {
      const tokens = await collectTokens(jest.fn());
      const result = await transformToScssWithInstances(tokens);

      // Should still contain instance mixin but with comment about no overrides
      expect(result.result).toContain('@mixin unchanged-button');
      expect(result.result).toContain('// No override styles - inherits all styles from component');
    } finally {
      figma.getNodeByIdAsync = originalGetNodeById;
    }
  });

  it('should create snapshot of generated SCSS with instances', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();
    global.figma = testSetup.figma;

    const originalGetNodeById = figma.getNodeByIdAsync;
    figma.getNodeByIdAsync = jest.fn().mockImplementation(async (id: string) => {
      const findNode = (nodes: any[], targetId: string): any => {
        for (const node of nodes) {
          if (node.id === targetId) return node;
          if (node.children) {
            const found = findNode(node.children, targetId);
            if (found) return found;
          }
        }
        return null;
      };
      return findNode([testData], id);
    });

    try {
      const tokens = await collectTokens(jest.fn());
      const result = await transformToScssWithInstances(tokens);

      // Use snapshot testing to ensure consistent output
      expect(result.result).toMatchSnapshot('scss-with-instances');
      expect(result.warnings).toMatchSnapshot('instance-warnings');
    } finally {
      figma.getNodeByIdAsync = originalGetNodeById;
    }
  });
});
