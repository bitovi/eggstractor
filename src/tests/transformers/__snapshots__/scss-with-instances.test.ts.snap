// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instance Override SCSS Generation should create snapshot of generated SCSS with instances: instance-warnings 1`] = `[]`;

exports[`Instance Override SCSS Generation should create snapshot of generated SCSS with instances: scss-with-instances 1`] = `
"
// Generated SCSS Mixins
"
`;
